File: Week_06/647/palindromic_substrings.py
class Solution:
    def countSubstrings(self, s: str) -> int:
        n = len(s)
        if n < 2:
            return n
        dp = [[False] * n for _ in range(n)]
        res = 0

        for i in range(n):
            for j in range(0, i + 1):
                if s[i] == s[j] and (i - j < 2 or dp[i - 1][j + 1]):
                    dp[i][j] = True
                    res += 1
        return res

File: Week_06/647/palindromic_substrings.go
package main

func countSubstrings(s string) int {
	n := len(s)
	if n < 2 {
		return n
	}
	dp := make([][]bool, n)
	for i := 0; i < n; i++ {
		dp[i] = make([]bool, n)
	}
	res := 0
	for i := 0; i < n; i++ {
		for j := 0; j <= i; j++ {
			if s[i] == s[j] && (i-j < 2 || dp[i-1][j+1]) {
				dp[i][j] = true
				res++
			}
		}
	}
	return res
}
